### PR TITLE
feat: allow captcha to be enabled when logging in

### PR DESCRIPTION
**Pull request by wenxuan70**
fix: #972

https://user-images.githubusercontent.com/55948263/194995734-fad86ef6-983a-4d98-93e0-7fc8be00bb95.mov

### CODE REVIEW

**Comment by wenxuan70** [comment]
fix: #897

https://user-images.githubusercontent.com/55948263/195017446-7e091f43-807f-45db-8d61-7b17b46d6c6e.mov

**Comment by wenxuan70** [comment]
https://user-images.githubusercontent.com/55948263/195623380-0d71fb28-6377-49d5-859b-3d1de816529f.mov

**Comment by wenxuan70** [comment]
An application has the following two options for human verification:
- `never`: not enabled
- `everytime`: human-machine verification is performed every time a user logs in with a password

File: web/src/CaptchaWaySelect.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
+// Copyright 2021 The Casdoor Authors. All Rights Reserved.
```
Comment (hsluoyz) [inline web/src/CaptchaWaySelect.js:1]: Remove this file

File: web/src/auth/LoginPage.js [hunk 2]
```diff
@@ -50,6 +51,7 @@ class LoginPage extends React.Component {
       validPhone: false,
       loginMethod: "password",
     };
+    this.captchaModal = React.createRef();
```
Comment (leo220yuyaodog) [inline web/src/auth/LoginPage.js:54]: Can not login in.

image

File: web/src/auth/LoginPage.js [hunk 3]
```diff
@@ -453,6 +478,33 @@ class LoginPage extends React.Component {
     }
   }
 
+  renderCaptchaModal(application) {
+    if (this.state.application && application?.providers?.length > 0) {
+      const provider = application.providers.filter(provider => {
+        return provider.provider?.category === "Captcha" && provider.provider?.type === "Default";
+      })[0];
+      // eslint-disable-next-line no-console
```
Comment (leo220yuyaodog) [inline web/src/auth/LoginPage.js:486]: delete

File: web/src/ProviderTable.js [hunk 4]
```diff
@@ -193,6 +193,26 @@ class ProviderTable extends React.Component {
...
+        title: i18next.t("application:rule"),
+        key: "rule",
+        width: "100px",
+        render: (text, record, index) => {
+          if (record.provider?.category !== "Captcha") {
+            return null;
+          }
+          return (
+            <Select virtual={false} style={{width: "100%"}}
+              value={record?.rule}
```
Comment (hsluoyz) [inline web/src/ProviderTable.js:206]: text

File: web/src/auth/LoginPage.js [hunk 5]
```diff
@@ -453,6 +478,33 @@ class LoginPage extends React.Component {
...
+      if (provider && provider.provider && provider?.rule === "Always") {
```
Comment (hsluoyz) [inline web/src/auth/LoginPage.js:487]: provider can be removed?

File: web/src/locales/de/data.json [hunk 6]
```diff
@@ -233,7 +233,9 @@
     "Users": "Benutzer",
     "Users under all organizations": "Benutzer unter allen Organisationen",
     "Webhooks": "Webhooks",
-    "{total} in total": "{total} insgesamt"
+    "{total} in total": "{total} insgesamt",
+    "Never": "Never",
+    "Everytime": "Everytime"
```
Comment (hsluoyz) [inline web/src/locales/de/data.json:238]: No this

**Review by hsluoyz — CHANGES_REQUESTED** [review]

File: web/src/auth/LoginPage.js (hunk 5 above)
Comment (leo220yuyaodog) [reply web/src/auth/LoginPage.js:487]: // eslint-disable-next-line no-console 
not provider

**Comment by wenxuan70** [comment]
@leo220yuyaodog Can you help me with code review? thanks

**Comment by leo220yuyaodog** [comment]
> @leo220yuyaodog Can you help me with code review? thanks

Ok

File: web/src/ProviderTable.js [hunk 7]
```diff
@@ -193,6 +193,28 @@ class ProviderTable extends React.Component {
       //     )
       //   }
       // },
+      {
+        title: i18next.t("application:rule"),
```
Comment (leo220yuyaodog) [inline web/src/ProviderTable.js:197]:
```suggestion
        title: i18next.t("application:Rule"),
```

**Review by leo220yuyaodog — CHANGES_REQUESTED** [review]
Generate i18 https://casdoor.org/docs/contributing#how-to-update-i18n-data

File: web/src/ProviderTable.js [hunk 8]
```diff
@@ -193,6 +193,28 @@ class ProviderTable extends React.Component {
...
+            return null;
+          }
+          return (
+            <Select virtual={false} style={{width: "100%"}}
+              value={text}
+              defaultValue="None"
+              onChange={value => {
+                this.updateField(table, index, "rule", value);
+              }} >
+              <Option key="None" value="None">None</Option>
```
Comment (leo220yuyaodog) [inline web/src/ProviderTable.js:212]:
```suggestion
              <Option key="None" value={i18next.t("application:None")}>None</Option>
```

File: web/src/ProviderTable.js [hunk 9]
```diff
@@ -193,6 +193,28 @@ class ProviderTable extends React.Component {
...
+              <Option key="Always" value="Always">Always</Option>
```
Comment (leo220yuyaodog) [inline web/src/ProviderTable.js:213]: i18

Comment (leo220yuyaodog) [inline web/src/ProviderTable.js:213]:
```suggestion
              <Option key="Always" value={i18next.t("application:Always")}>Always</Option>
```

**Review by leo220yuyaodog — CHANGES_REQUESTED** [review]

File: web/src/ProviderTable.js (hunk 8 above)
Comment (hsluoyz) [reply web/src/ProviderTable.js:212]: Wrong place? @leo220yuyaodog

File: web/src/ProviderTable.js (hunk 9 above)
Comment (hsluoyz) [reply web/src/ProviderTable.js:213]: Wrong place? @leo220yuyaodog

**Comment by hsluoyz** [comment]
@wenxuan70

File: web/src/ProviderTable.js (hunk 7 above)
Comment (wenxuan70) [reply web/src/ProviderTable.js:197]: @leo220yuyaodog I found that `application:rule` is also used in `SignupTable.js`, do I need to modify it together?

Comment (leo220yuyaodog) [reply web/src/ProviderTable.js:197]: > @leo220yuyaodog I found that `application:rule` is also used in `SignupTable.js`, do I need to modify it together?

Yes, modify together.

File: web/src/ProviderTable.js (hunk 9 above)
Comment (wenxuan70) [reply web/src/ProviderTable.js:213]: Does the text in this option need to be modified?

Comment (wenxuan70) [reply web/src/ProviderTable.js:213]: @hsluoyz

Comment (leo220yuyaodog) [reply web/src/ProviderTable.js:213]: > Does the text in this option need to be modified?

Yes, when switch the language, the options also need to be translated.

Comment (leo220yuyaodog) [reply web/src/ProviderTable.js:213]: > Wrong place? @leo220yuyaodog

I modified the code in wrong place.  Now it has be corrected

**Review by leo220yuyaodog — APPROVED** [review]

**Review by hsluoyz — APPROVED** [review]

**Pull request merged**